ranger-discovery-bundle/src/main/java/io/appform/ranger/discovery/bundle/id/DistributedIdGenerator.java

```diff
@@ -0,0 +1,288 @@
+                        .node(Integer.parseInt(matcher.group(3)))
+                        .exponent(Integer.parseInt(matcher.group(4)))
+                        .generatedDate(DATE_TIME_FORMATTER.parseDateTime(matcher.group(2)).toDate())
+                        .build());
+            }
+            return Optional.empty();
+        }
+        catch (Exception e) {
+            log.warn("Could not parse idString {}", e.getMessage());
+            return Optional.empty();
+        }
+    }
+
+    private int getTargetPartitionId() {
+        return SECURE_RANDOM.nextInt(partitionCount);
+    }
+
+    private Optional<Integer> getTargetPartitionId(final List<KeyValidationConstraint> inConstraints, final boolean skipGlobal) {
+//      ToDo: Check if we need Collision Checker here
+        return Optional.ofNullable(
+                RETRIER.get(() -> SECURE_RANDOM.nextInt(partitionCount)))
+                .filter(key -> validateId(inConstraints, key, skipGlobal));
+    }
+
+    protected boolean validateId(final List<KeyValidationConstraint> inConstraints,
+                                 final int partitionId,
+                                 final boolean skipGlobal) {
+        //First evaluate global constraints
+        val failedGlobalConstraint
+                = skipGlobal
+                  ? null
+                  : GLOBAL_CONSTRAINTS.stream()
+                          .filter(constraint -> !constraint.isValid(partitionId))
+                          .findFirst()
+                          .orElse(null);
+        if (null != failedGlobalConstraint) {
+            return false;
+        }
+        //Evaluate local + domain constraints
+        val failedLocalConstraint
+                = null == inConstraints
+                  ? null
+                  : inConstraints.stream()
+                          .filter(constraint -> !constraint.isValid(partitionId))
+                          .findFirst()
+                          .orElse(null);
+        return null == failedLocalConstraint;
+    }
+
+    protected int readRetryCount() {
+        try {
+            val count = Integer.parseInt(System.getenv().getOrDefault("NUM_ID_GENERATION_RETRIES", "512"));
+            if (count <= 0) {
+                throw new IllegalArgumentException(
+                        "Negative number of retries does not make sense. Please set a proper value for " +
+                                "NUM_ID_GENERATION_RETRIES");
+            }
+            return count;
+        }
+        catch (NumberFormatException e) {
+            throw new IllegalArgumentException("Please provide a valid positive integer for NUM_ID_GENERATION_RETRIES");
+        }
+    }
+
+    private void deleteExpiredKeys() {
+        val timeThreshold = DateTime.now().getMillis() / 1000;
+        for (val entry : dataStore.entrySet()) {
+            entry.getValue().entrySet().removeIf(partitionIdTrackerEntry -> partitionIdTrackerEntry.getKey() < timeThreshold);
```

Comment:
Need to think and handle cases where time changes on a machine, probably safer to remove entries which are timeThreshold - SOME_VALUE